### PR TITLE
Communication between the restful golang app and postgres db running …

### PR DESCRIPTION
**Pull request by ivshil**
Waiting for code review!
docker compose up
localhost:8000 for database adminer
localhost:1337 for app
localhost:1337/users get request to print the avaiable users

### CODE REVIEW

File: Dockerfile [hunk 1]
```diff
@@ -0,0 +1,24 @@
+# Use the official Go image as the base image
+FROM golang:latest
+
+# Set the working directory inside the container
```
Comment (gudata) [inline Dockerfile]: Comments are usually needed when you have to put some brain CPU to decode the next line.
In that case it is clear that those comments are chatgpt descriptions

File: Dockerfile [hunk 2]
```diff
@@ -0,0 +1,24 @@
...
+RUN go build -o main .
+
+# Expose the port on which the Go app will listen
+EXPOSE 1337
+
+# Set environment variables for the PostgreSQL connection
+ENV PGHOST=postgres
+ENV PGPORT=5432
+ENV PGUSER=admin
+ENV PGPASSWORD=url_short
```
Comment (gudata) [inline Dockerfile]: I know that this is dummy project but please do not store passwords in git.

This should be only

ENV PGPASSWORD

and the value must be provided externally - from .envrc file or the bash environment.

check and try to install https://sobolevn.me/git-secret/ if you want to store passwords in git.

File: Dockerfile [hunk 3]
```diff
@@ -0,0 +1,24 @@
...
+ENV PGDATABASE=postgres_url_short_db
```
Comment (gudata) [inline Dockerfile]: poor naming.
Please check if you wold like that one:

ENV POSTGRES_DATABASE=short_urls

File: docker-compose.yml [hunk 4]
```diff
@@ -0,0 +1,39 @@
+version: '3.9'
+
+services:
+  postgres:
+    image: postgres:latest
+    container_name: postgres_url_short
+    restart: unless-stopped
```
Comment (gudata) [inline docker-compose.yml:7]: this will auto start on your pc boot.

File: docker-compose.yml [hunk 5]
```diff
@@ -0,0 +1,39 @@
...
+    container_name: go_url_short
+    build:
+      context: .
+    restart: unless-stopped
+    depends_on:
+      - postgres
+    environment:
+      DATABASE_URL: postgres://admin:url_short@postgres/postgres_url_short_db
+    ports:
+      - "1337:1337"
```
Comment (gudata) [inline docker-compose.yml]: cool

File: main.go [hunk 6]
```diff
@@ -0,0 +1,252 @@
...
+
+	_ "github.com/lib/pq"
+)
+
+// Database connection parameters
+const (
+	dbHost = "postgres"
+	dbPort = 5432
+	dbUser = "admin"
+	dbPass = "url_short"
```
Comment (gudata) [inline main.go]: please load from config file or from environment.

File: main.go [hunk 7]
```diff
@@ -0,0 +1,252 @@
...
+		if err != nil {
+			http.Error(w, fmt.Sprintf("Error scanning users: %v", err), http.StatusInternalServerError)
+			return
+		}
+		users = append(users, user)
+	}
+
+	// Convert users slice to JSON and write the response
+	w.Header().Set("Content-Type", "application/json")
+	err = json.NewEncoder(w).Encode(users)
```
Comment (gudata) [inline main.go:177]: does it work with dates?

File: main.go [hunk 8]
```diff
@@ -0,0 +1,252 @@
...
+	}
+
+	var user User
+	if err := json.NewDecoder(r.Body).Decode(&user); err != nil {
+		http.Error(w, fmt.Sprintf("Error decoding JSON request: %v", err), http.StatusBadRequest)
+		return
+	}
+
+	query := "INSERT INTO users (name, email, birthdate) VALUES ($1, $2, $3) RETURNING user_id"
+	err := db.QueryRow(query, user.Name, user.Email, user.BirthDate).Scan(&user.ID)
```
Comment (gudata) [inline main.go:198]: nice - that's the right way to query 
that way you will avoid sql injection.

File: main.go [hunk 9]
```diff
@@ -0,0 +1,252 @@
...
+	w.Header().Set("Content-Type", "application/json")
+	err = json.NewEncoder(w).Encode(user)
+	if err != nil {
+		http.Error(w, fmt.Sprintf("Error encoding response to JSON: %v", err), http.StatusInternalServerError)
+		return
+	}
+}
+
+func getUrlShortsHandler(w http.ResponseWriter, r *http.Request, db *sql.DB) {
+	rows, err := db.Query("SELECT * FROM url_shorts")
```
Comment (gudata) [inline main.go:213]: good practice is tables to be plural.

File: main.go [hunk 10]
```diff
@@ -0,0 +1,252 @@
...
+	if err != nil {
+		http.Error(w, fmt.Sprintf("Error fetching url_shorts: %v", err), http.StatusInternalServerError)
+		return
+	}
+	defer rows.Close()
+
+	var urlShorts []URLShort
+	for rows.Next() {
```
Comment (gudata) [inline main.go:221]: If you extract the logic to get the shorts in a service you will be in a good position to write a nice unit test.

File: Dockerfile [hunk 11]
```diff
@@ -0,0 +1,24 @@
+# Use the official Go image as the base image
+FROM golang:latest
```
Comment (gudata) [inline Dockerfile]: going on production? - don't forget to hard code the version which is known to work with your code.
:latest is good for development and tests.

**Review by gudata — APPROVED** [review]
Nice style! Keep to good work!

File: Dockerfile (hunk 1 above)
Comment (ivshil) [reply Dockerfile]: ACK, Removed most of the useless comments

File: Dockerfile (hunk 2 above)
Comment (ivshil) [reply Dockerfile]: I moved everything sensitively to .env file, in the near future I'll try the git secret too

File: docker-compose.yml (hunk 4 above)
Comment (ivshil) [reply docker-compose.yml:7]: Doesn't seem to do that.

In summary, the restart: unless-stopped directive in the docker-compose.yml file ensures that the PostgreSQL container is automatically restarted by Docker if it exits unexpectedly but does not handle automatic startup on your PC boot. To achieve automatic startup on boot, you'll need to set up a system service to manage the container's lifecycle.

File: main.go (hunk 10 above)
Comment (ivshil) [reply main.go:221]: Maybe after adding all the basic functionality i will try to refactor the code in a proper structure, probably something like this:
myapp/
  ├── cmd/
  │   └── main.go
  ├── internal/
  │   ├── models/
  │   │   └── user.go
  │   ├── repository/
  │   │   └── user_repository.go
  │   ├── service/
  │   │   └── user_service.go
  │   └── handler/
  │       └── user_handler.go
  ├── pkg/
  │   └── database/
  │       └── database.go
  ├── migrations/
  │   └── <your_migration_files>.sql
  ├── config/
  │   └── config.yaml
  ├── Dockerfile
  └── docker-compose.yml

File: main.go [hunk 12]
```diff
@@ -22,26 +27,55 @@ var (
...
+func (cd CustomDate) Value() (driver.Value, error) {
+	return time.Time(cd), nil
+}
+
+func (cd CustomDate) MarshalJSON() ([]byte, error) {
+	t := time.Time(cd)
+	formattedDate := t.Format("2006-01-02")
+	return json.Marshal(formattedDate)
+}
+
```
Comment (gudata) [inline main.go:58]: Is this big enough to live in its model?

File: main.go [hunk 13]
```diff
@@ -119,18 +144,15 @@ func main() {
 		}
 	})
 
-	// Start the HTTP server
 	log.Printf("Server listening on port %s...\n", goAppPort)
 	log.Fatal(http.ListenAndServe(":"+goAppPort, nil))
 }
 
-// openDB opens a connection to the PostgreSQL database
```
Comment (gudata) [inline main.go]: Now the code is much readable without those redundant comments

File: main.go [hunk 14]
```diff
@@ -172,6 +194,7 @@ func addUserHandler(w http.ResponseWriter, r *http.Request, db *sql.DB) {
 	}
 
 	query := "INSERT INTO users (user_name, user_email, user_bdate) VALUES ($1, $2, $3) RETURNING user_id"
+	fmt.Printf(query, user.Name, user.Email, user.BirthDate)
```
Comment (gudata) [inline main.go:197]: Please start using the logger for those kind of activity. This should fit in the DEBUG log level.

File: main.go [hunk 15]
```diff
@@ -266,3 +289,94 @@ func getTaskContributorsHandler(w http.ResponseWriter, r *http.Request, db *sql.
...
+func urlFormHandler(w http.ResponseWriter, r *http.Request) {
+	if r.Method == http.MethodGet {
+		tmpl := template.Must(template.ParseFiles("url.html"))
+		tmpl.Execute(w, nil)
+	} else {
+		http.Error(w, "Method not allowed", http.StatusMethodNotAllowed)
+	}
+}
+
+func submitURLHandler(db *sql.DB) http.HandlerFunc {
```
Comment (gudata) [inline main.go:304]: This method is too long.
It contains also some logic which needs to be extracted in a model or a service.

File: main.go [hunk 16]
```diff
@@ -266,3 +289,94 @@ func getTaskContributorsHandler(w http.ResponseWriter, r *http.Request, db *sql.
...
+	return func(w http.ResponseWriter, r *http.Request) {
+		if r.Method == http.MethodPost {
+			r.ParseForm()
+			urlInput := r.Form.Get("url")
```
Comment (gudata) [inline main.go:308]: Please research the max length of a get request supported on browsers on linux/windows/macos and make sure you have good amount of space to store it.

I recommend hard coding a maximum size of the url and trim the rest.
if the len > SOME_CONST then warn the user that the url was trimmed.

File: main.go [hunk 17]
```diff
@@ -266,3 +289,94 @@ func getTaskContributorsHandler(w http.ResponseWriter, r *http.Request, db *sql.
...
+
+func isURLSExists(db *sql.DB, urls string) bool {
+	query := "SELECT COUNT(*) FROM url_shorts WHERE url_short = $1"
+	var count int
+	err := db.QueryRow(query, urls).Scan(&count)
+	if err != nil {
+		return false
+	}
+	return count > 0
+}
```
Comment (gudata) [inline main.go:366]: can those live in a model or service?

**Review by gudata — APPROVED** [review]

**Pull request merged**